solidity/contracts/middleware/liquidity-layer/adapters/CCTPAdapter.sol

```diff
@@ -0,0 +1,172 @@
+        address _token,
+        string calldata _bridge,
+        address _mailbox,
+        address _interchainGasPaymaster,
+        address _interchainSecurityModule
+    ) external initializer {
+        __HyperlaneConnectionClient_initialize(
+            _mailbox,
+            _interchainGasPaymaster,
+            _interchainSecurityModule,
+            _owner
+        );
+
+        tokenMessenger = ITokenMessenger(_tokenMessenger);
+        circleMessageTransmitter = ICircleMessageTransmitter(
+            _circleMessageTransmitter
+        );
+        token = _token;
+        bridge = _bridge;
+    }
+
+    function transferRemote(
+        uint32 _destinationDomain,
+        bytes32 _recipientAddress,
+        uint256 _amount
+    ) external override returns (bytes32) {
+        _mustHaveRemoteRouter(_destinationDomain);
+        uint32 _circleDomain = hyperlaneDomainToCircleDomain[
+            _destinationDomain
+        ];
+
+        // Approve the token to Circle. We assume that the LiquidityLayerRouterV2
+        // has already transferred the token to this contract.
+        require(
+            IERC20(token).approve(address(tokenMessenger), _amount),
+            "!approval"
+        );
+
+        uint64 _nonce = tokenMessenger.depositForBurn(
+            _amount,
+            _circleDomain,
+            _recipientAddress, // Mint to the recipient itself. This is different from the CircleBridgeAdapter.
+            token
+        );
+
+        emit BridgedToken(_nonce);
+
+        bytes memory _adapterData = abi.encode(_nonce, TOKEN_SYMBOL);
+        // The user's message "wrapped" required by this middleware
+        bytes memory _messageWithEmptyMetadata = abi.encode(
+            TypeCasts.addressToBytes32(msg.sender),
+            _recipientAddress, // The "user" recipient
+            _amount, // The amount of the tokens sent over the bridge
+            bridge, // The destination token bridge ID
+            _adapterData, // The adapter-specific data
+            bytes("") // Empty "user" message
+            // TODO : remove handling of user message in the router because it will only be handled by the ICARouter
+        );
+
+        // Dispatch the _messageWithEmptyMetadata to the destination's LiquidityLayerRouter.
+        return _dispatch(_destinationDomain, _messageWithEmptyMetadata);
+    }
+
+    // token transfer is already handled by the CCTPIsm
+    function _handle(
+        uint32, // origin
+        bytes32, // sender
+        bytes calldata // message
+    ) internal pure override {
+        // do nothing
+    }
+
+    function quoteGasPayment(uint32 _destinationDomain)
+        external
+        view
+        override(GasRouter, ILiquidityLayerAdapterV2)
+        returns (uint256 _gasPayment)
+    {
+        return
+            interchainGasPaymaster.quoteGasPayment(
+                _destinationDomain,
+                destinationGas[_destinationDomain]
+            );
+    }
+
+    function payForGas(
```

Comment:
Sorry this is my bad, but looking at `TokenRouter`, we actually pay for the gas inside the `transferRemote`, so we should probably adjust the`ILiquidityLayerAdapterV2` and remove `payForGas` from the interface?

Reply:
Done